graphql/src/execution/cache.rs

```diff
@@ -75,3 +78,162 @@ impl<R: CheapClone> QueryCache<R> {
         (work.await, cached)
     }
 }
+
+#[derive(Debug)]
+struct CacheByBlock {
+    block: EthereumBlockPointer,
+    max_weight: usize,
+    weight: usize,
+
+    // The value is `(result, n_hits)`.
+    cache: HashMap<QueryHash, (Arc<QueryResult>, AtomicU64)>,
+}
+
+impl CacheByBlock {
+    fn new(block: EthereumBlockPointer, max_weight: usize) -> Self {
+        CacheByBlock {
+            block,
+            max_weight,
+            weight: 0,
+            cache: HashMap::new(),
+        }
+    }
+
+    fn get(&self, key: &QueryHash) -> Option<&Arc<QueryResult>> {
+        let (value, hit_count) = self.cache.get(key)?;
+        hit_count.fetch_add(1, Ordering::SeqCst);
+        Some(value)
+    }
+
+    /// Returns `true` if the insert was successful or `false` if the cache was full.
+    fn insert(&mut self, key: QueryHash, value: Arc<QueryResult>, weight: usize) -> bool {
+        // We never try to insert errors into this cache, and always resolve some value.
+        assert!(!value.has_errors());
+        let fits_in_cache = self.weight + weight <= self.max_weight;
+        if fits_in_cache {
+            self.weight += weight;
+            self.cache.insert(key, (value, AtomicU64::new(0)));
```

Comment:
Should we also track how long inserts take for each `CacheByBlock`? Since it might be related to allocations, it would be good to instrument the `CacheByBlock` to keep a `total_insertion_time`